File: src/main/java/com/zippeykeys/praisebe/registry/DeityRegistry.java
package com.zippeykeys.praisebe.registry;

import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Stream;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.zippeykeys.praisebe.deity.Deity;
import com.zippeykeys.praisebe.util.Util;

import net.minecraftforge.registries.IForgeRegistry;


@Singleton
public class DeityRegistry extends PBRegistry<Deity>{
    @Inject
    @SuppressWarnings("deprecation")
    public DeityRegistry(IForgeRegistry<Deity> registryIn){
        super(Deity.Type.class, Deity.Element.class, Deity.Alignment.class);
        registryIn.getValues() //
          .forEach(this::register);
    }

    public Deity register(Deity value){
        return register(Util.getResource(value.getName()), value);
    }

    public Deity[] registerAll(Deity... values){
        return registerAll(Arrays.stream(values));
    }

    public Deity[] registerAll(Collection<? extends Deity> values){
        return registerAll(values.stream());
    }

    public Deity[] registerAll(Stream<? extends Deity> values){
        return values.map(this::register) //
          // .filter(Objects::nonNull) // Maybe shouldn't clean
          .toArray(Deity[]::new);
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/registry/DeityRegistry.kt
package com.zippeykeys.praisebe.registry

import com.zippeykeys.praisebe.deity.Deity
import com.zippeykeys.praisebe.util.getResource
import net.minecraftforge.registries.IForgeRegistry
import java.util.Arrays
import java.util.Objects
import java.util.stream.Stream
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class DeityRegistry @Inject constructor(registryIn: IForgeRegistry<Deity>) :
        PBRegistry<Deity>(Deity.Type::class.java, Deity.Element::class.java, Deity.Alignment::class.java) {
    init {
        registryIn.valuesCollection.forEach { this.register(it) }
    }

    fun register(value: Deity): Deity? {
        return set(getResource(value.name), value)
    }

    fun registerAll(vararg values: Deity): Array<Deity> {
        return registerAll(Arrays.stream(values))
    }

    fun registerAll(values: Collection<Deity>): Array<Deity> {
        return registerAll(values.stream())
    }

    @Suppress("UNCHECKED_CAST")
    fun registerAll(values: Stream<out Deity>): Array<Deity> {
        return values.map { register(it) }
                .filter(Objects::nonNull) // Maybe shouldn't clean
                .toArray() as Array<Deity>
    }
}